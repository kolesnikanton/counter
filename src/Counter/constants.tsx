export const ERROR_MESSAGES = {
  smallDurationValue: 'Too small duration value',
  emptyStartEndValue: 'Start and end props are empty',
  incorrectProps: 'Incorrect props',
};

export const MINIMAL_ANIMATION_INTERVAL_TIME = 500; // 0.5 sec
export const MINIMAL_INTERVAL_TIME = 10;
export const DEFAULT_VALUES = {
  fontSize: 80,
};
