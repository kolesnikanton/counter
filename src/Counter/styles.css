.counter-number {
  font-weight: bold;
  position: relative;
  overflow: hidden;
}

.next,
.current {
  position: absolute;
  transition-property: transform;
  transition-timing-function: linear;
}
