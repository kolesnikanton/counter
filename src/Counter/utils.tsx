export function isFloat(str: string): boolean {
  return str.includes('.');
}

export function getIntervalStepTime({
  duration, startNumber, endNumber, isDecrease, decimals,
}: {
  startNumber: string,
  endNumber: string;
  duration: number;
  isDecrease: boolean;
  decimals: number;
}): number {
  const start = Number(startNumber);
  const end = Number(endNumber);

  // Calculate how many operations we need to make before setInterval is finished
  const operations = isDecrease ? start - end : end - start;

  if (isFloat(endNumber) || isFloat(startNumber)) {
    // Multiply operations by number after point
    // 10 ** decimals - operations after point
    const factor = 10 ** decimals;
    const operationsWithFloat = operations * factor;
    return duration / operationsWithFloat;
  }

  return Math.floor(duration / operations);
}

export function isIntervalEnd({
  nextIntervalValue,
  end,
  decimals,
}: {
  nextIntervalValue: string;
  end: string;
  decimals: number;
}): boolean {
  // For a case where start = *.0 and end = *.01
  if (isFloat(nextIntervalValue) || isFloat(end)) {
    return nextIntervalValue === Number(end).toFixed(decimals);
  }

  return nextIntervalValue === end;
}

export function getDecimals({ start, end }: {
  start: string,
  end: string,
}): number {
  const startDecimals = isFloat(start) ? start.split('.')[1].length : 0;
  const endDecimals = isFloat(end) ? end.split('.')[1].length : 0;

  return startDecimals > endDecimals ? startDecimals : endDecimals;
}

export function getNextIntervalValue({
  isFloatRange, currentValue, decimals, operation,
}: {
  isFloatRange: boolean,
  currentValue: number,
  decimals: number,
  operation: number,
  // factor: number,
}): string {
  if (isFloatRange) {
    const factor = 10 ** decimals;
    // operation / factor = 0.1, 0.001, 0,0001... (Depends on decimals)
    const result = (currentValue + operation / factor).toFixed(decimals);
    return result;
  }

  const result = currentValue + operation;
  return String(result);
}
