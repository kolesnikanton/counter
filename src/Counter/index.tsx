import React, { useRef, useEffect } from 'react';

import Error from './Error';
import { isInt, getIntervalStepTime, getNextIntervalValue } from './utils';
import { ERROR_MESSAGES, SMALL_DURATION_VALUE, DEFAULT_VALUES } from './constants';

/* eslint-disable no-param-reassign */
export default function Counter({
  duration,
  start = 0,
  end = 0,
  className,
  withAnimation,
  fontSize = DEFAULT_VALUES.fontSize,
}: {
  duration: number,
  start?: number,
  end?: number,
  className?: string,
  withAnimation?: boolean,
  fontSize?: number,
}) {
  const inputRef = useRef<HTMLDivElement>(null);
  const isDecrease = start > end;
  const currentInputValueRef = useRef('');

  function setInputPosition({
    isInitial,
    intervalStepTime,
    currentEl,
    nextEl,
  }: {
    isInitial: boolean;
    intervalStepTime: number;
    currentEl: HTMLDivElement;
    nextEl: HTMLDivElement;
  }) {
    const transition = isInitial ? 'transform 0s' : `transform ${intervalStepTime / 2}ms`;
    currentEl.style.transition = transition;
    nextEl.style.transition = transition;

    const trasnlateValue = fontSize;
    const currentElTranslate = isInitial
      ? 'translate(0, 0)'
      : `translate(0, ${isDecrease ? trasnlateValue : -trasnlateValue}px)`;

    const nextElTranslate = isInitial
      ? `translate(0, ${isDecrease ? -trasnlateValue : trasnlateValue}px)`
      : 'translate(0, 0)';

    currentEl.style.transform = currentElTranslate;
    nextEl.style.transform = nextElTranslate;
  }

  function setInitialValue(value: string, currentEl: HTMLDivElement) {
    currentInputValueRef.current = value;
    currentEl.innerText = value;
  }

  function changeInputValue(
    nextInputValue: string,
    intervalStepTime: number,
    isFloatRange: boolean,
    currentEl: HTMLDivElement,
    nextEl: HTMLDivElement,
  ) {
    if (isFloatRange || !withAnimation) {
      currentEl.innerText = nextInputValue;
      return;
    }

    setInputPosition({
      isInitial: true,
      intervalStepTime,
      currentEl,
      nextEl,
    });
    currentEl.innerText = currentInputValueRef.current;
    nextEl.innerText = nextInputValue;

    setTimeout(() => {
      setInputPosition({
        isInitial: false,
        intervalStepTime,
        currentEl,
        nextEl,
      });
    }, intervalStepTime / 2);

    currentInputValueRef.current = nextInputValue;
  }

  function count({ intervalStepTime, counterEl }: {
    intervalStepTime: number;
    counterEl: HTMLDivElement;
  }) {
    const currentEl = counterEl.children[0] as HTMLDivElement;
    const nextEl = counterEl.children[1] as HTMLDivElement;

    setInitialValue(String(start), currentEl);

    const isFloatRange = !isInt(end) || !isInt(start);
    let currentIntervalValue = start;

    const timer = setInterval(() => {
      const nextIntervalValue = getNextIntervalValue({
        isFloatRange,
        isDecrease,
        currentValue: currentIntervalValue,
      });

      changeInputValue(
        isFloatRange ? nextIntervalValue.toFixed(1) : String(nextIntervalValue),
        intervalStepTime,
        isFloatRange,
        currentEl,
        nextEl,
      );

      if (nextIntervalValue === end) {
        clearInterval(timer);
      }

      currentIntervalValue = nextIntervalValue;
    }, intervalStepTime);
  }

  if (!start && !end) {
    return <Error errorMessage={ERROR_MESSAGES.emptyStartEndValue} />;
  }

  const intervalStepTime = getIntervalStepTime({
    startNumber: start, endNumber: end, duration, isDecrease,
  });

  if (!intervalStepTime) {
    return <Error errorMessage={ERROR_MESSAGES.smallDurationValue} />;
  }

  if (withAnimation && intervalStepTime < SMALL_DURATION_VALUE) {
    return <Error errorMessage={ERROR_MESSAGES.smallDurationValue} />;
  }

  useEffect(() => {
    if (!inputRef.current) {
      return;
    }

    count({
      intervalStepTime,
      counterEl: inputRef.current,
    });
  }, []);

  const defaultClassName = 'counter-number';
  const classNames = className ? `${defaultClassName} ${className}` : defaultClassName;
  const startLength = String(start).length;
  const endLength = String(end).length;
  const maxLength = startLength > endLength ? startLength : endLength;
  const fontWidth = fontSize * 0.6;
  const fontHeight = fontSize * 1.15;
  const width = maxLength * fontWidth;

  return (
    <div
      ref={inputRef}
      className={classNames}
      style={{ fontSize, height: fontHeight, width }}
    >
      <div className="current" />
      <div className="next" />
    </div>
  );
}
