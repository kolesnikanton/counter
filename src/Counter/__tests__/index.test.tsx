import React from 'react';
import { render } from '@testing-library/react';
import Counter from '..';

test('render', async () => {
  jest.useFakeTimers();

  const { container } = render(
    <Counter fontSize={50} start="2" end="23.5" duration={10000} />,
  );

  const input1 = container.getElementsByClassName('current')[0] as HTMLDivElement;
  expect(input1.innerText).toBe('2');

  jest.runAllTimers();

  const input2 = container.getElementsByClassName('current')[0] as HTMLDivElement;
  expect(input2.innerText).toBe('23.5');

  const nextEl = container.getElementsByClassName('next')[0] as HTMLDivElement;
  expect(nextEl.innerText).toBe(undefined);

  expect(container).toMatchSnapshot();
});

test('render - withAnimation', async () => {
  jest.useFakeTimers();

  const { container } = render(
    <Counter withAnimation fontSize={80} start="200" end="100" duration={100000} />,
  );

  const input1 = container.getElementsByClassName('current')[0] as HTMLDivElement;
  expect(input1.innerText).toBe('200');

  jest.runAllTimers();

  const nextEl = container.getElementsByClassName('next')[0] as HTMLDivElement;
  expect(nextEl.innerText).toBe('100');

  const currentEl = container.getElementsByClassName('current')[0] as HTMLDivElement;
  expect(currentEl.innerText).toBe('101');

  expect(container).toMatchSnapshot();
});

test('render with floats', async () => {
  jest.useFakeTimers();

  const { container } = render(
    <Counter fontSize={80} start="10.1" end="100.10" duration={100000} />,
  );

  const input1 = container.getElementsByClassName('current')[0] as HTMLDivElement;
  expect(input1.innerText).toBe('10.1');

  jest.runAllTimers();

  const input2 = container.getElementsByClassName('current')[0] as HTMLDivElement;
  expect(input2.innerText).toBe('100.10');

  const nextEl = container.getElementsByClassName('next')[0] as HTMLDivElement;
  expect(nextEl.innerText).toBe(undefined);

  expect(container).toMatchSnapshot();
});

test('render with floats with decimals', async () => {
  jest.useFakeTimers();

  const { container } = render(
    <Counter decimals={2} fontSize={80} start="10.10000" end="100.10000000" duration={100000} />,
  );

  const input1 = container.getElementsByClassName('current')[0] as HTMLDivElement;
  expect(input1.innerText).toBe('10.10000');

  jest.runAllTimers();

  const input2 = container.getElementsByClassName('current')[0] as HTMLDivElement;
  expect(input2.innerText).toBe('100.10');

  const nextEl = container.getElementsByClassName('next')[0] as HTMLDivElement;
  expect(nextEl.innerText).toBe(undefined);

  expect(container).toMatchSnapshot();
});

test('render without start', async () => {
  jest.useFakeTimers();

  const { container } = render(
    <Counter fontSize={80} end="100" duration={100000} />,
  );

  const input1 = container.getElementsByClassName('current')[0] as HTMLDivElement;
  expect(input1.innerText).toBe('0');

  jest.runAllTimers();

  const input2 = container.getElementsByClassName('current')[0] as HTMLDivElement;
  expect(input2.innerText).toBe('100');

  const nextEl = container.getElementsByClassName('next')[0] as HTMLDivElement;
  expect(nextEl.innerText).toBe(undefined);

  expect(container).toMatchSnapshot();
});

test('render - duration error', async () => {
  const { container } = render(
    <Counter fontSize={80} start="101" end="100" duration={1} />,
  );

  expect(container).toMatchSnapshot();
});

test('render - without numbers error', async () => {
  const { container } = render(
    <Counter start="" end="" fontSize={80} duration={100} />,
  );

  expect(container).toMatchSnapshot();
});
