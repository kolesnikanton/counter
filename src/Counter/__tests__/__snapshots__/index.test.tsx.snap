// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render - duration error 1`] = `
<div>
  <span>
    Error: Too small duration value
  </span>
</div>
`;

exports[`render - withAnimation 1`] = `
<div>
  <div
    class="counter-number"
    style="font-size: 80px; height: 92px;"
  >
    <div
      class="current"
      style="height: 92px; transition: transform 500ms; transform: translate(0, 92px);"
    />
    <div
      class="next"
      style="height: 92px; transition: transform 500ms; transform: translate(0, -92px);"
    />
  </div>
</div>
`;

exports[`render - without numbers error 1`] = `
<div>
  <span>
    Error: Start and end props are empty
  </span>
</div>
`;

exports[`render 1`] = `
<div>
  <div
    class="counter-number"
    style="font-size: 50px; height: 57.49999999999999px;"
  >
    <div
      class="current"
      style="height: 57.49999999999999px;"
    />
    <div
      class="next"
      style="height: 57.49999999999999px;"
    />
  </div>
</div>
`;

exports[`render with floats 1`] = `
<div>
  <div
    class="counter-number"
    style="font-size: 80px; height: 92px;"
  >
    <div
      class="current"
      style="height: 92px;"
    />
    <div
      class="next"
      style="height: 92px;"
    />
  </div>
</div>
`;

exports[`render with floats with decimals 1`] = `
<div>
  <div
    class="counter-number"
    style="font-size: 80px; height: 92px;"
  >
    <div
      class="current"
      style="height: 92px;"
    />
    <div
      class="next"
      style="height: 92px;"
    />
  </div>
</div>
`;

exports[`render without start 1`] = `
<div>
  <div
    class="counter-number"
    style="font-size: 80px; height: 92px;"
  >
    <div
      class="current"
      style="height: 92px;"
    />
    <div
      class="next"
      style="height: 92px;"
    />
  </div>
</div>
`;
